tools/myst/host/utils.c

```diff
@@ -349,3 +350,148 @@ void free_mount_mapping_opts(myst_mount_mapping_t* mappings)
         free(mappings->mounts);
     }
 }
+
+int create_roothash_file(const char* path, char filename[PATH_MAX])
+{
+    int ret = 0;
+    myst_fssig_t fssig;
+    char ascii[MYST_SHA256_ASCII_SIZE];
+    int fd = -1;
+    char tmpfile[] = "/tmp/mystXXXXXX";
+
+    if (!path || !filename)
+    {
+        ret = -EINVAL;
+        goto done;
+    }
+
+    /* load the signature structure (fssig) from the rootfs if an ext2 */
+    if (myst_load_fssig(path, &fssig) != 0)
+    {
+        ret = -ENOTSUP;
+        goto done;
+    }
+
+    if (myst_bin_to_ascii(
+            fssig.root_hash, sizeof(fssig.root_hash), ascii, sizeof(ascii)) !=
+        0)
+    {
+        ret = -EINVAL;
+        goto done;
+    }
+
+    if ((fd = mkstemp(tmpfile)) < 0)
+    {
+        ret = -EINVAL;
+        goto done;
+    }
+
+    if (write(fd, ascii, sizeof(ascii) - 1) != sizeof(ascii) - 1)
+    {
+        ret = -EINVAL;
+        goto done;
+    }
+
+    myst_strlcpy(filename, tmpfile, PATH_MAX);
+
+done:
+
+    if (fd > 0)
+        close(fd);
+
+    return ret;
+}
+
+ssize_t create_roothashes(
+    const char* rootfs,
+    const myst_mount_mapping_t* mount_mapping,
+    char** roothashes,
+    size_t roothashes_size)
+{
+    ssize_t ret = 0;
+    size_t num_roothashes = 0;
+
+    /* if an ext2 rootfs, then create roothash file from rootfs */
+    {
+        char filename[PATH_MAX];
+        int ret = create_roothash_file(rootfs, filename);
+
+        if (ret == 0)
+        {
+            char* str;
+
+            if (num_roothashes == roothashes_size)
+            {
+                ret = -ERANGE;
+                goto done;
+            }
+
+            if (!(str = strdup(filename)))
+            {
+                ret = -ENOMEM;
+                goto done;
+            }
+
+            roothashes[num_roothashes++] = str;
+        }
+        else if (ret < 0 && ret != -ENOTSUP)
+        {
+            ret = -EINVAL;
+            goto done;
+        }
+    }
+
+    /* create roothash files for any mount mappings that are ext2 */
+    {
+        const size_t count = mount_mapping->mounts_count;
+        char** mounts = mount_mapping->mounts;
+
+        for (size_t i = 0; i < count; i++)
+        {
+            char path[2 * PATH_MAX + 1];
+            char filename[PATH_MAX];
+            char* eq;
+
+            myst_strlcpy(path, mounts[i], sizeof(path));
+
+            if (!(eq = strchr(path, '=')))
+            {
+                ret = -EINVAL;
+                goto done;
+            }
+
+            *eq = '\0';
+
+            int ret = create_roothash_file(path, filename);
```

Comment:
It's not clear to me why the mounted hostfs path would contain a roothash section? Why do we need to create root hash for mounted host ext2 fs? Is it verified during mounting or file accessing?

Reply:
The roothash region (part of the measured image) contains the roothashes that are trusted. When an app attempts to mount an ext2 file system, the ext2 image itself contains a self-proclaimed roothash (in the fssig section). The mounter verifies that the self-proclaimed roothash is trusted (i.e., that is is contained in the roothash region). Then during mounting, the hash tree is recomputed to ensure that the roothash is correct.